File: leetcode/198.house-robber.java
import java.util.Arrays;

/*
 * @lc app=leetcode id=198 lang=java
 *
 * [198] House Robber
 */

// @lc code=start
class Solution {
    public int rob(int[] nums) {
        // dp[i]=max(dp[i-1],dp[i-2]+nums[i]);
        if (nums.length < 2) {
            return nums[0];
        }
        int[] dp = new int[nums.length];
        dp[0] = nums[0];
        dp[1] = Math.max(dp[0], nums[1]);

        for (int i = 2; i < nums.length; ++i) {

            dp[i] = Math.max(dp[i - 1], dp[i - 2] + nums[i]);

        }
        System.out.println(Arrays.toString(dp));

        return dp[nums.length - 1];
    }
}
// @lc code=end

File: rust/leetcode/src/medium/_198_house_robber.rs
//DP
struct Solution;

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut dp = vec![0; n + 2];
        if n == 0 {
            return 0;
        }
        if n == 1 {
            return nums[0];
        }
        dp[1] = nums[0];
        dp[2] = nums[0].max(nums[1]);

        for i in 3..=n {
            dp[i] = dp[i - 1].max(dp[i - 2] + nums[i - 1]);
        }

        dp[n]
    }
}

#[test]
fn test() {
    assert_eq!(Solution::rob(vec![1, 2, 3, 1]), 4);
    assert_eq!(Solution::rob(vec![2, 7, 9, 3, 1]), 12);
}
